File: src/main/java/shateq/moonlight/dispatcher/Order.java
package shateq.moonlight.dispatcher;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Command meta data
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface Order {
    // Command's name
    String value();

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    @interface Rank {
        Category value() default Category.Blank;
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    @interface Aliases {
        String[] value();
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    @interface Explanation {
        String value();
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    @interface Example {
        String value();
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    @interface Hidden {
    }
}

File: src/main/java/shateq/moonlight/dispatcher/Order.kt
package shateq.moonlight.dispatcher

/**
 * Command meta data
 */
@Target(AnnotationTarget.TYPE)
@MustBeDocumented
annotation class Order(val value: String) {
    
    @Target(AnnotationTarget.TYPE)
    annotation class Rank(val value: Category = Category.General)

    @Target(AnnotationTarget.TYPE)
    annotation class Aliases(val value: Array<String>)

    @Target(AnnotationTarget.TYPE)
    annotation class Explanation(val value: String)

    @Target(AnnotationTarget.TYPE)
    annotation class Example(val value: String)

    @Target(AnnotationTarget.TYPE)
    annotation class Hidden
}
